pkg/api/controller_test.go

```diff
@@ -3860,64 +3860,157 @@ func TestController_CopyObjectHandler(t *testing.T) {
 	clt, deps := setupClientWithAdmin(t)
 	ctx := context.Background()
 
-	_, err := deps.catalog.CreateRepository(ctx, "repo1", onBlock(deps, "bucket/prefix"), "main")
-	if err != nil {
-		t.Fatal(err)
+	repo := testUniqueRepoName()
+	_, err := deps.catalog.CreateRepository(ctx, repo, onBlock(deps, "bucket/prefix"), "main")
+	require.NoError(t, err)
+	_, err = deps.catalog.CreateBranch(ctx, repo, "alt", "main")
+	require.NoError(t, err)
+
+	uploadContent := func(t *testing.T, repository, branch, objPath string) api.ObjectStats {
+		t.Helper()
+		const content = "hello world this is my awesome content"
+		uploadResp, err := uploadObjectHelper(t, ctx, clt, objPath, strings.NewReader(content), repository, branch)
+		verifyResponseOK(t, uploadResp, err)
+		require.NotNil(t, uploadResp.JSON201)
+		require.Equal(t, len(content), int(api.Int64Value(uploadResp.JSON201.SizeBytes)))
+		return *uploadResp.JSON201
 	}
 
-	const expectedSize = 38
-	const content = "hello world this is my awesome content"
-	uploadResp, err := uploadObjectHelper(t, ctx, clt, "foo/bar", strings.NewReader(content), "repo1", "main")
-	verifyResponseOK(t, uploadResp, err)
+	const copyTypeHeaderName = "X-Lakefs-Copy-Type"
+
+	t.Run("shallow", func(t *testing.T) {
+		const (
+			srcPath  = "foo/bar"
+			destPath = "foo/bar-shallow-copy"
+		)
+		objStat := uploadContent(t, repo, "main", srcPath)
+		copyResp, err := clt.CopyObjectWithResponse(ctx, repo, "main", &api.CopyObjectParams{
+			DestPath: destPath,
+		}, api.CopyObjectJSONRequestBody{
+			SrcPath: srcPath,
+		})
+		verifyResponseOK(t, copyResp, err)
 
-	sizeBytes := api.Int64Value(uploadResp.JSON201.SizeBytes)
+		copyType := copyResp.HTTPResponse.Header.Get(copyTypeHeaderName)
+		require.Equal(t, copyType, "shallow")
+		// Verify creation path, date and physical address are different
+		copyStat := copyResp.JSON201
+		require.NotNil(t, copyStat)
+		require.Equal(t, objStat.PhysicalAddress, copyStat.PhysicalAddress)
+		require.GreaterOrEqual(t, objStat.Mtime, copyStat.Mtime)
+		require.Equal(t, destPath, copyStat.Path)
 
-	if sizeBytes != expectedSize {
-		t.Fatalf("expected %d bytes to be written, got back %d", expectedSize, sizeBytes)
-	}
+		// Verify all else is equal
+		objStat.Mtime = copyStat.Mtime
+		objStat.Path = copyStat.Path
+		require.Nil(t, deep.Equal(objStat, *copyStat))
 
-	t.Run("copy object", func(t *testing.T) {
-		time.Sleep(1 * time.Second)
-		copyResp, err := clt.CopyObjectWithResponse(ctx, "repo1", "main", &api.CopyObjectParams{
-			DestPath: "bar/foo",
+		// get back info
+		statResp, err := clt.StatObjectWithResponse(ctx, repo, "main", &api.StatObjectParams{Path: destPath})
+		verifyResponseOK(t, statResp, err)
+		require.Nil(t, deep.Equal(statResp.JSON200, copyStat))
+	})
+
+	t.Run("full_diff_branch", func(t *testing.T) {
```

Comment:
diff has many meanings
```suggestion
	t.Run("full_different_branch", func(t *testing.T) {
```